### PR TITLE
use onnx "onehot" to map tf "onehot" if possible

### PR DESCRIPTION
**Pull request by zhijxu-MS**

### CODE REVIEW

File: tf2onnx/tfonnx.py [hunk 1]
```diff
@@ -1255,6 +1255,45 @@ def onehot_op(ctx, node, name, args):
         return
 
 
+def onehot_op9(ctx, node, name, args):
+    # T output = OneHot(uint8/int32/int64 input, T depth, T on-value, T off-value, @int axis, @dtype)
```
Comment (nbcsm) [inline tf2onnx/tfonnx.py]: let's follow onnx spec by default, assume all types are supported.
for now, we can make use of target rs6 for onnxruntime as a workaround...

**Comment by nbcsm** [comment]
https://github.com/onnx/tensorflow-onnx/issues/270

File: tf2onnx/tfonnx.py [hunk 2]
```diff
@@ -1255,6 +1255,45 @@ def onehot_op(ctx, node, name, args):
...
+    # tf requires that dtype is same as on-value's and off-value's dtype
+    # in ONNX, op's schema is (input, depth, [off-value, on-value], @int axis)
```
Comment (pengwa) [inline tf2onnx/tfonnx.py]: you can probably write that to "T3 output = OneHot(T1 indices, T2 depth, T3 values, @int axis)  # values=[off_value on_value]"

File: tf2onnx/tfonnx.py [hunk 3]
```diff
@@ -1255,6 +1255,45 @@ def onehot_op(ctx, node, name, args):
...
+    # onnxruntime only supports int64
+    output_dtype = ctx.get_dtype(node.input[2])
+    if output_dtype not in [onnx_pb.TensorProto.INT64, onnx_pb.TensorProto.INT32]:
+        log.warning("unsupported dtype in onnxruntime, onehot can't be used directly")
```
Comment (pengwa) [inline tf2onnx/tfonnx.py]: we should avoid new logic specifically for onnxruntime. If onnxruntime does not support, we have centralized place to do the cast. e.g. rs6.

File: tf2onnx/tfonnx.py (hunk 2 above)
Comment (zhijxu-MS) [reply tf2onnx/tfonnx.py]: done

File: tf2onnx/tfonnx.py (hunk 3 above)
Comment (zhijxu-MS) [reply tf2onnx/tfonnx.py]: talked with guohua, I will add check before adding cast node.

File: tf2onnx/tfonnx.py (hunk 1 above)
Comment (zhijxu-MS) [reply tf2onnx/tfonnx.py]: done

File: tests/test_backend.py [hunk 4]
```diff
@@ -1140,6 +1140,31 @@ def test_onehot2(self):
...
+        for np_dtype, tf_dtype in zip([np.int32, np.int64], [tf.int32, tf.int64]):
+            tf.reset_default_graph()
+            x_val = np.array([0, 1, 2, 1, 2, 0, 1, 2, 1, 2], dtype=np_dtype)
+            depth = np.array(20).astype(np.int64)
+            x = tf.placeholder(tf_dtype, x_val.shape, name=_TFINPUT)
+            on_off = np.array([5.6, 1.2]).astype(np_dtype)
+            x_ = tf.one_hot(x, depth, on_value=on_off[0], axis=-1, off_value=on_off[1])
+            _ = tf.identity(x_, name=_TFOUTPUT)
+            graph = self._run_test_case([_OUTPUT], {_INPUT: x_val})
+            self.assertTrue(len(group_nodes_by_type(graph)["OneHot"]) == 1, "onnx onehot should be used")
```
Comment (nbcsm) [inline tests/test_backend.py]: [optional] consider use https://github.com/onnx/tensorflow-onnx/blob/5feb8e8d42d3fc010020741a630fcc955c07d2d9/tests/common.py#L237-L238

File: tf2onnx/tfonnx.py [hunk 5]
```diff
@@ -1287,6 +1287,45 @@ def onehot_op(ctx, node, name, args):
...
+    depth = ctx.make_node("Unsqueeze", [depth], attr={"axes": [0]}).output[0]
+
+    on_value = node.input[2]
+    off_value = node.input[3]
+    on_value = ctx.make_node("Unsqueeze", [on_value], attr={"axes": [0]}).output[0]
+    off_value = ctx.make_node("Unsqueeze", [off_value], attr={"axes": [0]}).output[0]
+    off_on_value = ctx.make_node("Concat", [off_value, on_value], attr={"axis": 0}).output[0]
+
+    indices = node.input[0]
+    if ctx.get_dtype(indices) != onnx_pb.TensorProto.INT64 and ctx.is_target(TARGET_RS6):
```
Comment (nbcsm) [inline tf2onnx/tfonnx.py]: check ```ctx.is_target(TARGET_RS6)``` first, then type.

File: tf2onnx/tfonnx.py [hunk 6]
```diff
@@ -1287,6 +1287,45 @@ def onehot_op(ctx, node, name, args):
...
 
 
+def onehot_op9(ctx, node, name, args):
+    # T output = OneHot(uint8/int32/int64 input, T depth, T on-value, T off-value, @int axis, @dtype)
+    # tf requires that dtype is same as on-value's and off-value's dtype
+    # in ONNX, op's schema is (input, depth, value, @int axis), meaning of "value" is [off-value, on-value]
+    # onnxruntime only supports int64
+    output_dtype = ctx.get_dtype(node.input[2])
+    if ctx.is_target(TARGET_RS6) and output_dtype not in [onnx_pb.TensorProto.INT64, onnx_pb.TensorProto.INT32]:
+        log.warning("unsupported dtype in onnxruntime, onehot can't be used directly")
```
Comment (nbcsm) [inline tf2onnx/tfonnx.py]: please include node name in the message, and change onehot => onehot-**9**

File: tf2onnx/tfonnx.py [hunk 7]
```diff
@@ -1287,6 +1287,45 @@ def onehot_op(ctx, node, name, args):
...
+        onehot_op(ctx, node, name, args)
+
+    depth = node.input[1]
+    depth = ctx.make_node("Unsqueeze", [depth], attr={"axes": [0]}).output[0]
+
+    on_value = node.input[2]
+    off_value = node.input[3]
+    on_value = ctx.make_node("Unsqueeze", [on_value], attr={"axes": [0]}).output[0]
+    off_value = ctx.make_node("Unsqueeze", [off_value], attr={"axes": [0]}).output[0]
+    off_on_value = ctx.make_node("Concat", [off_value, on_value], attr={"axis": 0}).output[0]
```
Comment (nbcsm) [inline tf2onnx/tfonnx.py]: are all these Unsqueeze aboved required?

File: tests/test_backend.py [hunk 8]
```diff
@@ -1140,6 +1140,31 @@ def test_onehot2(self):
         _ = tf.identity(x_, name=_TFOUTPUT)
```
Comment (nbcsm) [inline tests/test_backend.py]: add a message for skip test_onehot1 (above)

File: tests/test_backend.py [hunk 9]
```diff
@@ -1140,6 +1140,31 @@ def test_onehot2(self):
...
         self._run_test_case([_OUTPUT], {_INPUT: x_val})
 
+    @check_opset_min_version(9, "onehot")
+    def test_onehot3(self):
```
Comment (nbcsm) [inline tests/test_backend.py]: please also add a few cases to cover different axis.

Comment (nbcsm) [reply tests/test_backend.py]: also add an axis=1 case for test_onehot2

File: tf2onnx/tfonnx.py (hunk 7 above)
Comment (zhijxu-MS) [reply tf2onnx/tfonnx.py]: concat requires vector

**Pull request merged**